elements/dash-tree/dash-tree.js

```diff
@@ -0,0 +1,405 @@
+      // keys, home or end. The element calls `preventDefault` to prevent the
+      // browser from taking any actions.
+      event.preventDefault();
+    }
+
+    /**
+     * Return the current activedescendant if there is one. Otherwise,
+     * return null.
+     */
+    _currentTreeItem() {
+      const activeDescendant = this.getAttribute('aria-activeDescendant');
+      if (activeDescendant) {
+        return this.querySelector(`#${activeDescendant}`);
+      }
+
+      return null;
+    }
+
+    /**
+     * `_nextTreeItem` gets the treeitem that comes after the currently active
+     * one.
+     */
+    _focusNextTreeItem() {
+      const treeItems = this._allTreeItems();
+      const currentItem = this._currentTreeItem();
+
+      let next = undefined;
+      for (let i = treeItems.length - 1; i >= 0; i--) {
+        let treeItem = treeItems[i];
+        if (treeItem === currentItem) {
+          break;
+        }
+        if (treeItem.visible) {
+          next = treeItem;
+        }
+      }
+      if (next)
+        this._focusTreeItem(next);
+    }
+
+    /**
+     * `_prevTreeItem` gets the treeitem that comes before the currently
+     * selected one.
+     */
+    _focusPrevTreeItem() {
+      const treeItems = this._allTreeItems();
+      const currentItem = this._currentTreeItem();
+
+      let prev = undefined;
+      for (let i = 0; i < treeItems.length; i++) {
+        let treeItem = treeItems[i];
+        if (treeItem === currentItem) {
+          break;
+        }
+        if (treeItem.visible) {
+          prev = treeItem;
+        }
+      }
+      if (prev)
+        this._focusTreeItem(prev);
+    }
+
+    /**
+     * Focus the first `<dash-treeitem>` in the tree.
+     */
+    _focusFirstTreeItem() {
+      this._focusTreeItem(this.querySelector('dash-treeitem'));
+    }
+
+    /**
+     * Focus the last `<dash-treeitem>` in the tree that is focusable without
+     * opening a node.
+     */
+    _focusLastTreeItem() {
+      const treeItems = this._allTreeItems();
+      for (let i = treeItems.length - 1; i >= 0; i--) {
+        let treeItem = treeItems[i];
+        if (treeItem.visible) {
+          this._focusTreeItem(treeItem);
+          break;
+        }
+      }
+    }
+
+    /**
+     * `_focusTreeItem` marks the given treeitem as the new activedescendant.
+     */
+    _focusTreeItem(treeItem) {
+      this.setAttribute('aria-activedescendant', treeItem.id);
+
+      // There can be only one active item at a time.
+      // Find any previous active item and remove its active class.
+      const activeItem = this.querySelector('.active');
+      if (activeItem)
+        activeItem.classList.remove('active');
+      // If the element is expandable then it should have a <span> label
+      // as an immediate child. Put the `active` class on that child instead.
+      if (treeItem.hasAttribute('aria-expanded')) {
+        treeItem.firstElementChild.classList.add('active');
```

Comment:
Since the `<span>` is semantically a label, I’d require a `<span>` on leafs _and_ inner nodes. The distinction of leafs and inner nodes happens with a contained `<dash-itemgroup>`. You can add the `<span>` automatically if it’s missing maybe. WDYT?